woocommerce/woocommerce-gutenberg-products-block
# WIP Postcode validation/character limit

- Converted: `BillingCountryInput`, `ShippingCountryInput`, `CountryInput`, `BillingStateInput`, `ShippingStateInput`, `StateInput`, and `AddressForm` to TypeScript.
- Port [`is_postcode`](https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-wc-validation.php#L47) from WC Core to JavaScript, created a wrapper function for it called `validatePostcode` and added unit tests, also ported from WC Core.
- Added a new (optional) prop to `ValidatedTextInput` called `customValidation`. It is a function that takes a string as its input, and returns a boolean, representing the validity of the field's value. You will see that I have added a check to see if the field being rendered is the postcode field, and if it is, include a custom validation function which calls `validatePostcode`.
- Moved some variables around in `ValidatedTextInput` to get access to `errorMessage` earlier. This is needed so we can set the validation error message to be the value of the field's `errorMessage`. Previously it looks like this was never used.
- Check to see if the `ValidatedTextInput`s prop `customValidation` is a function. If it is, call it on blur.


Fixes #3326




#### Accessibility




### Screenshots



### How to test the changes in this Pull Request:

1. Add items to the cart and go to the checkout page.
2. Enter an invalid postcode and verify an error shows.
3. Change country and enter another invalid postcode - ensure error still shows.
3. Enter a valid postcode and ensure error disappears.
4. Change country and enter a valid postcode, ensure no error is shown.
5. General smoke test on checkout.



### Changelog

> Add suggested changelog entry here.
